### PR TITLE
[webui + api]: reducing memory footprint and complexity of models and controllers

### PR DESCRIPTION
**Pull request by vpereira**
generally avoiding temporary arrays, removing unecessary returns and getting rid of if-else-if-else blocks

### CODE REVIEW

File: src/api/app/jobs/project_status_calculator.rb [hunk 1]
```diff
@@ -211,7 +212,7 @@ def calc_status(opts = {})
 
     links = Array.new
     # find links
-    mypackages.values.each do |package|
+    mypackages.each_value.each do |package|
       if package.project == @dbproj.name and package.links_to_id
```
Comment (coolo) [inline src/api/app/jobs/project_status_calculator.rb]: that doesn't make sense to me

File: src/api/app/jobs/project_status_calculator.rb [hunk 2]
```diff
@@ -131,26 +129,29 @@ def parse_jobhistory(dname, repo, arch)
...
-    unless d.blank?
-      data = Xmlhash.parse(d)
 
-      data.elements('jobhist') do |p|
+    return [] if d.blank?
+
+    data = Xmlhash.parse(d)
+
+    data.elements('jobhist').map do |p|
         line = {'name' => p['package'],
```
Comment (coolo) [inline src/api/app/jobs/project_status_calculator.rb]: very careful here - there is a difference if you can elements with a block or as such - as you talk about memory usage

Comment (coolo) [inline src/api/app/jobs/project_status_calculator.rb]: especially as this code has about 0 test coverage ;(

File: src/api/app/jobs/project_status_calculator.rb (hunk 1 above)
Comment (vpereira) [inline src/api/app/jobs/project_status_calculator.rb]: if you do .values, you create an array, that isn't necessary. Better to use an iterator...

File: src/api/app/jobs/project_status_calculator.rb (hunk 2 above)
Comment (vpereira) [inline src/api/app/jobs/project_status_calculator.rb]: for the GC, its easier to clean it up than ```` results << p ```

File: src/api/app/jobs/project_status_calculator.rb (hunk 1 above)
Comment (vpereira) [inline src/api/app/jobs/project_status_calculator.rb]: https://gist.github.com/vpereira/d3f9df7039df23f72f61

File: src/api/app/jobs/project_status_calculator.rb (hunk 2 above)
Comment (coolo) [inline src/api/app/jobs/project_status_calculator.rb]: not sure what line you're commenting, but in the lines above my comment is no results

File: src/api/app/jobs/project_status_calculator.rb (hunk 1 above)
Comment (coolo) [inline src/api/app/jobs/project_status_calculator.rb]: and if you do each_value.each you create an array too and then an iterator on top of it

Comment (vpereira) [inline src/api/app/jobs/project_status_calculator.rb]: ah now I see. the each shouldnt be there ;)

File: src/api/app/jobs/project_status_calculator.rb (hunk 2 above)
Comment (vpereira) [inline src/api/app/jobs/project_status_calculator.rb]: im commenting about the original code. It was the pattern:

```
results = []
data.elements('jobhist').each do |p|
  ....
  results.push(p)
end
```

Comment (coolo) [inline src/api/app/jobs/project_status_calculator.rb]: the code you're patching is
``
data.elements('jobhist') do |p|
``

Comment (vpereira) [inline src/api/app/jobs/project_status_calculator.rb]: true! however it was more to remove the ret variable... any other idea/suggestion?

**Comment by vpereira** [comment]
adding tests are actually decreasing the test coverage. Are the tests as well being counted as *untested* code?

**Comment by hennevogel** [comment]
Click [Details](https://coveralls.io/builds/2077232) to find out :-)

**Comment by vpereira** [comment]
maybe you have some different profile, I don't have a details to click on. But my last commit added just test cases, how did it reduce the coverage?

**Comment by hennevogel** [comment]
You're not reading this on github.com right? :) See https://coveralls.io/builds/2077232

**Comment by vpereira** [comment]
hehe for sure I'm reading on coveralls.io, but I cannot see for example a report as generated by simplecov. Looking the coverage generated by simplecov, it isn't decreasing -.-

**Comment by hennevogel** [comment]
:+1:

**Comment by adrianschroeter** [comment]
all, but the "[webapi] added test cases for refactored code" one got merged. thanks!

**Comment by vpereira** [comment]
ok  thanks!
 @adrianschroeter In the src/api/app/controllers/status_controller.rb, method history, I just tried to remove the being/rescue from the logic, since it's extreme slow http://technicaldebt.com/the-cost-of-using-rubys-rescue-as-logic/

**Comment by adrianschroeter** [comment]
we did not use any exception rescue there, they are just conditional returns... so I do not understand why the other code should be faster.

**Comment by vpereira** [comment]
i think yes, in the way that you are assigning hours and samples, no?

**Comment by adrianschroeter** [comment]
@vpereira I want to close this issue, do you want to rework the remaining commit first or shall we close it now?

**Comment by vpereira** [comment]
Hi @adrianschroeter  I don't have nothing to change here. Except by the the project name parsing using case/when (and documenting the regexp) - that I agree that its matter of taste, I believe all other changes are good.. 

the construction:

    params[:samples].respond_to?(:to_i) ?  params[:samples].to_i : 0

IMO is better than 

     samples = begin
       Integer(params[:samples] || '100') rescue 0
     end

feel free to drop it but please remove the  Array.new,Hash.new and "return"  calls  tha were "fixed" in the same commit... 

thanks!

**Pull request closed without merging**